Don't allocate backing stores for non-animated compositing layers with zero opacity

mattwoodrow opened the pull request:
#### 6d4c5f059b8767e4b3a7696a94a90660bb411048
<pre>
Don&apos;t allocate backing stores for non-animated compositing layers with zero opacity
<a href="https://bugs.webkit.org/show_bug.cgi?id=241935">https://bugs.webkit.org/show_bug.cgi?id=241935</a>

Reviewed by Simon Fraser.

We currently allocate (and paint) a backing store for opacity:0 layers, so that we can
initiate animations faster. Unfortunately this uses a lot of memory, so we&apos;re going to
try skipping this, and allocating/painting on demand when an animation is started.

This adds a new test with various opacity:0 layers, with and without animations. It also
tests the case where we animate 0 to 0, which could in theory skip a backing store, but
doesn&apos;t yet (and is unlikely to be a problem in the wild).

* LayoutTests/compositing/backing/zero-opacity-expected.txt: Added.
* LayoutTests/compositing/backing/zero-opacity.html: Added.
* LayoutTests/compositing/backing/zero-opacity-invalidation-expected.txt: Added.
* LayoutTests/compositing/backing/zero-opacity-invalidation.html: Added.
* LayoutTests/compositing/geometry/bounds-ignores-hidden-dynamic-negzindex-expected.txt:
* LayoutTests/compositing/geometry/fixed-position-flipped-writing-mode-expected.txt:
* LayoutTests/compositing/visibility/layer-visible-content-expected.txt:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::containsPaintedContent const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::mayHaveNonZeroOpacity const):
* Source/WebCore/style/Styleable.h:

Canonical link: <a href="https://commits.webkit.org/251965@main">https://commits.webkit.org/251965@main</a>
</pre>

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -2873,6 +2873,11 @@ bool RenderLayerBacking::containsPaintedContent(PaintedContentsInfo& contentsInf
     if (contentsInfo.isDirectlyCompositedImage())
         return false;
 
+    if (auto styleable = Styleable::fromRenderer(renderer())) {
```
What's the invalidation code path that runs when opacity changes from zero to non-zero or vice versa?   `RenderElement::adjustStyleDifference()` will see a chance for `StyleDifferenceContextSensitiveProperty::Opacity` but I think we're missing something that triggers this code running.

smfr replied:
I think you should also add a test for simple changes of opacity between 0 and non-zero.

alanbaradlay commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1049,6 +1049,10 @@ static bool rareNonInheritedDataChangeRequiresLayerRepaint(const StyleRareNonInh
 
     if (first.opacity != second.opacity) {
+        // If we're changing to/from 0 opacity, then repaint the layer since we try to skip buffer allocation
+        // for opacity:0.
```
I think this needs a bit more explanation as paint invalidation and buffer allocation are highly unrelated at this level.

mattwoodrow replied:
Is there a better solution that lets us do this invalidation in a location that makes more sense?

I could add a new StyleDifferenceContextSensitiveProperty enum value, and then do the check in adjustStyleDifference?

smfr replied:
You need to do something that will result in `RenderLayer::setNeedsCompositingConfigurationUpdate()` getting called, probably via `RenderLayerCompositor::layerStyleChanged()`, so I'd expect a change between zero and non-zero opacity to result in a `StyleDifference::RecompositeLayer`.

mattwoodrow committed:
Don't allocate backing stores for non-animated compositing layers with zero opacity
https://bugs.webkit.org/show_bug.cgi?id=241935

Reviewed by Simon Fraser.

We currently allocate (and paint) a backing store for opacity:0 layers, so that we can
initiate animations faster. Unfortunately this uses a lot of memory, so we're going to
try skipping this, and allocating/painting on demand when an animation is started.

This adds a new test with various opacity:0 layers, with and without animations. It also
tests the case where we animate 0 to 0, which could in theory skip a backing store, but
doesn't yet (and is unlikely to be a problem in the wild).

* LayoutTests/compositing/backing/zero-opacity-expected.txt: Added.
* LayoutTests/compositing/backing/zero-opacity.html: Added.
* LayoutTests/compositing/backing/zero-opacity-invalidation-expected.txt: Added.
* LayoutTests/compositing/backing/zero-opacity-invalidation.html: Added.
* LayoutTests/compositing/geometry/bounds-ignores-hidden-dynamic-negzindex-expected.txt:
* LayoutTests/compositing/geometry/fixed-position-flipped-writing-mode-expected.txt:
* LayoutTests/compositing/visibility/layer-visible-content-expected.txt:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::containsPaintedContent const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::mayHaveNonZeroOpacity const):
* Source/WebCore/style/Styleable.h:

Canonical link: https://commits.webkit.org/251965@main
